File: swift/Sources/AttomusOTP/HOTP.swift
#if canImport(CryptoKit)
import CryptoKit
#else
import Crypto
#endif
import Foundation

public enum HOTP {
    public static func generate(
        secret: Data,
        counter: UInt64,
        algorithm: OTPAlgorithm = .sha1,
        digits: Int = 6
    ) throws -> String {
        try OTPValidation.validateHOTPSecret(secret)
        try OTPValidation.validateHOTPDigits(digits)

        let counterData = counter.bigEndianData
        let hmac = try authenticationCode(for: counterData, secret: secret, algorithm: algorithm)
        let truncated = truncate(hmac)
        let modulus = UInt32(pow(10.0, Double(digits)))
        let code = truncated % modulus
        return String(format: "%0\(digits)u", code)
    }

    private static func authenticationCode(
        for counterData: Data,
        secret: Data,
        algorithm: OTPAlgorithm
    ) throws -> Data {
        let key = SymmetricKey(data: secret)

        switch algorithm {
        case .sha1:
            return Data(HMAC<Insecure.SHA1>.authenticationCode(for: counterData, using: key))
        case .sha256:
            return Data(HMAC<SHA256>.authenticationCode(for: counterData, using: key))
        case .sha512:
            return Data(HMAC<SHA512>.authenticationCode(for: counterData, using: key))
        }
    }

    private static func truncate(_ hmac: Data) -> UInt32 {
        let bytes = Array(hmac)
        let offset = Int(bytes[bytes.count - 1] & 0x0F)

        let part0 = UInt32(bytes[offset] & 0x7F) << 24
        let part1 = UInt32(bytes[offset + 1]) << 16
        let part2 = UInt32(bytes[offset + 2]) << 8
        let part3 = UInt32(bytes[offset + 3])

        return part0 | part1 | part2 | part3
    }
}

private extension UInt64 {
    var bigEndianData: Data {
        withUnsafeBytes(of: bigEndian) { Data($0) }
    }
}

File: kotlin/src/main/kotlin/com/attomus/otp/HOTP.kt
package com.attomus.otp

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object HOTP {
    fun generate(
        secret: ByteArray,
        counter: Long,
        algorithm: OTPAlgorithm = OTPAlgorithm.SHA1,
        digits: Int = 6
    ): String {
        OTPValidation.validateHOTPSecret(secret)
        OTPValidation.validateHOTPDigits(digits)

        val counterBytes = counter.toBigEndianBytes()
        var hmac: ByteArray? = null

        try {
            hmac = authenticationCode(counterBytes, secret, algorithm)
            val truncated = truncate(hmac)
            val modulus = modulusForDigits(digits)
            val code = truncated % modulus
            return code.toString().padStart(digits, '0')
        } finally {
            counterBytes.zero()
            hmac?.zero()
        }
    }

    private fun authenticationCode(
        counterBytes: ByteArray,
        secret: ByteArray,
        algorithm: OTPAlgorithm
    ): ByteArray {
        val macAlgorithm = when (algorithm) {
            OTPAlgorithm.SHA1 -> "HmacSHA1"
            OTPAlgorithm.SHA256 -> "HmacSHA256"
            OTPAlgorithm.SHA512 -> "HmacSHA512"
        }

        val mac = Mac.getInstance(macAlgorithm)
        mac.init(SecretKeySpec(secret, macAlgorithm))
        return mac.doFinal(counterBytes)
    }

    private fun truncate(hmac: ByteArray): Long {
        val offset = hmac[hmac.lastIndex].toInt() and 0x0f
        val part0 = (hmac[offset].toInt() and 0x7f) shl 24
        val part1 = (hmac[offset + 1].toInt() and 0xff) shl 16
        val part2 = (hmac[offset + 2].toInt() and 0xff) shl 8
        val part3 = hmac[offset + 3].toInt() and 0xff
        return (part0 or part1 or part2 or part3).toLong()
    }

    private fun modulusForDigits(digits: Int): Long {
        return when (digits) {
            6 -> 1_000_000L
            7 -> 10_000_000L
            8 -> 100_000_000L
            else -> error("Unexpected digit count")
        }
    }

    private fun Long.toBigEndianBytes(): ByteArray {
        val bytes = ByteArray(8)
        var value = this
        for (index in 7 downTo 0) {
            bytes[index] = (value and 0xffL).toByte()
            value = value ushr 8
        }
        return bytes
    }

    private fun ByteArray.zero() {
        fill(0)
    }
}

